src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/TokenData.cs

```diff
@@ -63,50 +64,48 @@ public int CompareTo(TokenData other)
         Contract.ThrowIfFalse(this.TokenStream == other.TokenStream);
 
         if (this.IndexInStream >= 0 && other.IndexInStream >= 0)
-        {
             return this.IndexInStream - other.IndexInStream;
-        }
+
+        if (this.Token == other.Token)
+            return 0;
 
         var start = this.Token.SpanStart - other.Token.SpanStart;
         if (start != 0)
-        {
             return start;
-        }
 
         var end = this.Token.Span.End - other.Token.Span.End;
         if (end != 0)
-        {
             return end;
-        }
 
-        // this is expansive check. but there is no other way to check.
+        // We have two different tokens, which are at the same location.  This can happen with things like empty/missing
+        // tokens.  In order to give a strict ordering, we need to walk up the tree to find the first common ancestor
+        // and see which token we hit first in that ancestor.
         var commonRoot = this.Token.GetCommonRoot(other.Token);
         RoslynDebug.Assert(commonRoot != null);
 
-        var tokens = commonRoot.DescendantTokens();
+        using var _ = ArrayBuilder<SyntaxNodeOrToken>.GetInstance(out var stack);
```

Comment:
>using var _ = ArrayBuilder<SyntaxNodeOrToken>.GetInstance(out var stack); [](http://example.com/codeflow?start=8&length=73)

Maybe it would be more complex to write, but it feels like walking up from token/other.Token and until right before hitting commonRoot will yield the children that need to be compared. Then it's just a matter of finding which ones child index is earlier, right?

Reply:
Thanks.  That idea worked nicely.